app/server/lib/uploads.ts

```diff
@@ -116,6 +118,69 @@ export async function handleUpload(req: Request, res: Response): Promise<UploadR
   return upload;
 }
 
+export interface MultipartFormFile {
+  name: string;
+  contentType: string;
+  stream: stream.Readable;
+}
+
+/**
+ * Processes a request containing a multipart/form-data body.
+ * @param {e.Request} req - Request to read body from
+ * @param {(file: MultipartFormFile) => Promise<void>} onFile
+ *  Called for each file found. The returned promise should only resolve
+ *  when the handler is finished with the data stream, otherwise errors might occur.
+ * @param {(name: string, value: string) => void} onField
+ *  Called for each field found.
+ * @returns {Promise<void>}
+ *  Promise, resolves when all parts of the form have been handled, or an error has occurred.
+ */
+export async function parseMultipartFormRequest(
```

Comment:
I want to test this, but I can't find a way of mocking Request that we already have in the codebase.

- Do we have a way to mock requests?
- If not, should I add a package like node-mocks-http?
- If not, is testing via DocAPI enough?

Reply:
As in spin up a HTTP server, expose this as an endpoint, then make fetch requests to it?

Reply:
Oh, this particular function doesn't have an exposed endpoint, is that the problem? Otherwise we could use [the usual test server.](https://github.com/gristlabs/grist-core/blob/a2489e0812fc0bc61955ab09c666cc185ced8406/test/gen-server/apiUtils.ts#L27)

Personally, it feels easier to start the server and use `fetch` or `axios.get` than to write mocks. What sort of tests did you have in mind?

Reply:
General behaviour tests for this function, things like
- It waits for all part handlers to be processed before resolving
- It errors on a bad request
- It processes all parts of a multi-part form
- It doesn't pass 'fields' to the .on('part') handler
- It rejects on a connection issue
- It passes the file details correctly to the handler.

Anything to do with making sure the function interacts with the underlying `multiparty` library correctly, really.

Reply:
And yeah - no exposed endpoint here. We do test the archive endpoint which uses it, but that doesn't cover all edge cases.